core/Segment.php

```diff
@@ -460,8 +465,46 @@ public function getHash()
 
     public static function getSegmentHash($definition)
     {
-        // urldecode to normalize the string, as browsers may send slightly different payloads for the same archive
-        return md5(urldecode($definition));
+        $cache = Cache::getEagerCache();
+        $cacheKey = self::CACHE_KEY . md5($definition);
+
+        if ($cache->contains($cacheKey)) {
+            return $cache->fetch($cacheKey);
+        }
+
+        $defaultHash = md5(urldecode($definition));
+
+        // if the cache for segments already built, but this segment was not found,
+        // we return the default segment, this can be a segment from url or
+        // something like "visitorType==new"
+        if ($cache->contains(self::SEGMENT_HAS_BUILT_CACHE_KEY)) {
+            return $defaultHash;
+        }
+
+        // the segment hash is not built yet, let's do it
+        $model = new SegmentEditorModel();
+        $segments = $model->getAllSegmentsAndIgnoreVisibility();
+
+        foreach ($segments as $segment) {
+            $cacheKeyTemp = self::CACHE_KEY . md5($segment['definition']);
+            $cache->save($cacheKeyTemp, $segment['hash']);
+
+            $cacheKeyTemp = self::CACHE_KEY . md5(urldecode($segment['definition']));
+            $cache->save($cacheKeyTemp, $segment['hash']);
+
+            $cacheKeyTemp = self::CACHE_KEY . md5(urlencode($segment['definition']));
+            $cache->save($cacheKeyTemp, $segment['hash']);
```

Comment:
Is it possible to write a test to ensure this logic doesn't result in collisions between different segments? For example the segments:

```
pageUrl==http://abc.com/d+f
pageUrl==http://abc.com/d f
pageUrl==http://abc.com/d%20f
```

and maybe also including operator characters in segment values like:

```
pageUrl==http://abc.com/=/@/!/,/;
```

(segments not encoded correctly for readability)

Reply:
ok, I'm on it

Reply:
@diosmosis 
with the current logic, the hash for these definitions is the same, because we `urldecode` the definitions before we `md5` them when we store them in the DB. 

```php
    private function createHash($definition)
    {
        return md5(urldecode($definition));
    }
```

Is that something we don't want?

Reply:
I added new tests @diosmosis